### PR TITLE
Create Keywords.md

### LINKED ISSUE #341: Add a page (or pages) with a list of keywords, and links to documentation
Such a list would make it easier for users to get and retain a fuller grasp of Typescript, by having an overview of the syntax of the language, and by clearly differentiating the parts that are common to ES5/6/7 and the parts that are unique to Typescript.

Presenting this information in this way, would allow users to reduce the question of "What does this keyword do in Typescript?", to "What does this keyword do differently in Typescript than in Javascript?".

Something like the Javascript keyword list on [MDN](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Lexical_grammar#Keywords) but also including a small one-line description of each keyword, similar to the [Javascript statment list on MSDN](https://msdn.microsoft.com/en-us/library/3xcfcb93(v=vs.94).aspx):

Keyword | Definition
--------|---------------
**const** | [Force the use of const enums](https://www.typescriptlang.org/docs/handbook/enums.html)

For the usages which Typescript shares with Javascript, I'm not sure which would be better -- not to mention them, or to simply include references to authoritative sources of Javascript documentation (MDN and/or MSDN):

Keyword | Language | Definition
---|---|---
**const** | JS | [MDN](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/const)
 | TS | [Force the use of const enums](https://www.typescriptlang.org/docs/handbook/enums.html)
**var** | JS | [MDN](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/var) [MSDN] (https://msdn.microsoft.com/en-us/library/z16cackw(v=vs.94).aspx)

### PR DESCRIPTION
**Pull request by zspitz**
Fixes #341 

cc @mhegazy

### CODE REVIEW

**Comment by zspitz** [comment]
Microsoft/Typescript#2536

File: pages/Keywords.md [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+Unless otherwise noted, more information can be found on the [Mozilla Developer Network](https://developer.mozilla.org/en-US/docs/Web/JavaScript) or on [MSDN](https://msdn.microsoft.com/en-us/library/d1et7k7c%28v=vs.94%29.aspx)
+
+---
+
+The following keywords are used for [basic type annotations](http://www.typescriptlang.org/docs/handbook/basic-types.html):
+
+Keyword | Description
+---|---
+any | [Describes a type unknown at design time](http://www.typescriptlang.org/docs/handbook/basic-types.html#any)
+boolean | [Boolean](http://www.typescriptlang.org/docs/handbook/basic-types.html#boolean)
```
Comment (mhegazy) [inline pages/Keywords.md]: use lower case type names, so as not to mix them with boxed types, Number, String, etc..

Comment (mhegazy) [inline pages/Keywords.md]: also add `never`.

File: pages/Keywords.md [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+null | _Missing Handbook link_ ([Spec](https://github.com/Microsoft/TypeScript/blob/master/doc/spec.md#326-the-null-type))
+number | [Number](http://www.typescriptlang.org/docs/handbook/basic-types.html#number)
+string | [String](http://www.typescriptlang.org/docs/handbook/basic-types.html#string)
+symbol | [Symbol](http://www.typescriptlang.org/docs/handbook/symbols.html)
+undefined | _Missing Handbook link_ ([Spec](https://github.com/Microsoft/TypeScript/blob/master/doc/spec.md#327-the-undefined-type))
+void | _Missing Handbook link_ ([Spec](https://github.com/Microsoft/TypeScript/blob/master/doc/spec.md#325-the-void-type))
```
Comment (mhegazy) [inline pages/Keywords.md]: we should also add `this`. as it has a special meaning.

File: pages/Keywords.md [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+
+---
+
+The following keywords are used for [basic type annotations](http://www.typescriptlang.org/docs/handbook/basic-types.html):
+
+Keyword | Description
+---|---
+any | [Describes a type unknown at design time](http://www.typescriptlang.org/docs/handbook/basic-types.html#any)
+boolean | [Boolean](http://www.typescriptlang.org/docs/handbook/basic-types.html#boolean)
+null | _Missing Handbook link_ ([Spec](https://github.com/Microsoft/TypeScript/blob/master/doc/spec.md#326-the-null-type))
```
Comment (mhegazy) [inline pages/Keywords.md]: https://github.com/Microsoft/TypeScript-Handbook/blob/master/pages/Basic%20Types.md#null-and-undefined

File: pages/Keywords.md [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+number | [Number](http://www.typescriptlang.org/docs/handbook/basic-types.html#number)
```
Comment (mhegazy) [inline pages/Keywords.md]: make links relative to this one. the contents of the handbook should be navigatable as md files on github. the script the publishes them to the website takes care of fixing the links.

File: pages/Keywords.md [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+string | [String](http://www.typescriptlang.org/docs/handbook/basic-types.html#string)
+symbol | [Symbol](http://www.typescriptlang.org/docs/handbook/symbols.html)
+undefined | _Missing Handbook link_ ([Spec](https://github.com/Microsoft/TypeScript/blob/master/doc/spec.md#327-the-undefined-type))
```
Comment (mhegazy) [inline pages/Keywords.md]: https://github.com/Microsoft/TypeScript-Handbook/blob/master/pages/Basic%20Types.md#null-and-undefined

File: pages/Keywords.md (hunk 2 above)
Comment (mhegazy) [inline pages/Keywords.md]: https://github.com/Microsoft/TypeScript-Handbook/blob/master/pages/Basic%20Types.md#void

File: pages/Keywords.md [hunk 6]
```diff
@@ -0,0 +1,97 @@
...
+void | _Missing Handbook link_ ([Spec](https://github.com/Microsoft/TypeScript/blob/master/doc/spec.md#325-the-void-type))
+
+Other basic type-related keywords:
+
+Keyword | Description
+---|---
+as | [Type assertion](http://www.typescriptlang.org/docs/handbook/basic-types.html#type-assetions)
```
Comment (mhegazy) [inline pages/Keywords.md]: also used as a module import rename declaration.

File: pages/Keywords.md [hunk 7]
```diff
@@ -0,0 +1,97 @@
...
+is | _Missing link_
```
Comment (mhegazy) [inline pages/Keywords.md]: see https://github.com/Microsoft/TypeScript-Handbook/blob/28108e225b85aa29666256c0db9a4a61b271bbe8/pages/Advanced%20Types.md#user-defined-type-guards

File: pages/Keywords.md [hunk 8]
```diff
@@ -0,0 +1,97 @@
...
+
+User-defined types:
+
+Keyword | Description
+---|---
+enum | [Defines an enum - a set of named values](http://www.typescriptlang.org/docs/handbook/enums.html)
+type | [Type alias](http://www.typescriptlang.org/docs/handbook/advanced-types.html#type-aliases)
+interface | [Defines a type by its shape (structural typing)](http://www.typescriptlang.org/docs/handbook/interfaces.html)
```
Comment (mhegazy) [inline pages/Keywords.md]: why not class too?

**Comment by mhegazy** [comment]
thanks @zspitz. since this is intended to be an index of sorts, i would recommend merging them all in one table. left column  has one key word, all sorted alphabetically of course. the right column can have multiple links based on the context. for the JS keywords, i would also add links to MDN documentation or ES6 spec.

**Comment by zspitz** [comment]
@mhegazy I'm thinking that an alphabetical arrangement might be important on paper, but within the browser there's not much benefit -- it's always possible to do CTRL+F to find a specific keyword on the page. However, isn't there added value in the grouping - type annotations, type modifiers, member modifiers etc.?
(I'll take care of the rest shortly.)

**Comment by mhegazy** [comment]
Well.. then what is the value of this page in general. you can always search this repo for keywords and that should get you close. like `class` would get you all the relevant areas. I was thinking of this more of a keyword index to get you quickly where you want, so you can look up `as`, and know it is used in 1. ES6 module declaration or 2. a post fix type assertion operator; similarly `async` would get you to some documentation about async functions in general.

**Comment by zspitz** [comment]
@mhegazy 
> you can always search this repo for keywords and that should get you close. like class would get you all the relevant areas.

I think many people would be coming from the Handbook link on the Typescript website, and not through the repo, thus repo search can't be counted on. Unless the plan is to eventually have that link refer to the repo (similar to the spec), and not to http://www.typescriptlang.org/docs?

> so you can look up `as`, and know it is used in ...
True. But perhaps sometimes you want to look up the available usages (basic type, type modifier, property modifier etc.) and quickly find out from there which keywords can fill each role. (I'm thinking computed enums here...). Admittedly, the hard part will be to figure out the different usage groups first.

File: pages/Keywords.md (hunk 2 above)
Comment (zspitz) [inline pages/Keywords.md]: Fixed `void`

>`this` as it has a special meaning

Does Typescript add more than the Javascript meaning of global context / call dependent context?

File: pages/Keywords.md (hunk 1 above)
Comment (zspitz) [inline pages/Keywords.md]: Fixed lowercase and `never`

File: pages/Keywords.md (hunk 3 above)
Comment (zspitz) [inline pages/Keywords.md]: Fixed

File: pages/Keywords.md (hunk 4 above)
Comment (zspitz) [inline pages/Keywords.md]: Fixed

File: pages/Keywords.md (hunk 5 above)
Comment (zspitz) [inline pages/Keywords.md]: Fixed

File: pages/Keywords.md (hunk 7 above)
Comment (zspitz) [inline pages/Keywords.md]: Fixed

File: pages/Keywords.md [hunk 9]
```diff
@@ -0,0 +1,99 @@
...
+namespace | [Associates the contained types with the specified namespace](namespaces.md)
+require | [Missing description](modules.md#export--and-import--require)
+
+---
+
+Unknown:
+
+Keyword |
+---|
+of|
```
Comment (DanielRosenwasser) [inline pages/Keywords.md]: `for-of`

Comment (zspitz) [inline pages/Keywords.md]: @DanielRosenwasser fixed

File: pages/Keywords.md (hunk 6 above)
Comment (zspitz) [inline pages/Keywords.md]: Fixed

**Comment by zspitz** [comment]
@mhegazy @DanielRosenwasser I've reorganized the keyword list in simple alphabetical order, and added links as appropriate (Handbook `||` spec `||` MDN)

1. The _this_-type referred to [here](https://github.com/Microsoft/TypeScript/blob/master/doc/spec.md#363-this-types) isn't a keyword; it refers to the type, but the `this` in code refers to the instance of the class. Other than that, `this` by itself has no added meaning in Typescript over Javascript; except for this-based type guards, which I noted.

2. There's an inconsistency between `if...else` which doesn't have a separate entry for `else`, and `as`, `from`, and `extends` which even though they can't stand alone, each have their own entry. The difference is between Javascript constructs and Typescript constructs.

**Comment by zspitz** [comment]
@mhegazy  @DanielRosenwasser I think this is pretty much complete. Could you look over it and see if I've missed anything? Any other thoughts?

File: pages/Keywords.md [hunk 10]
```diff
@@ -0,0 +1,173 @@
...
+| `as` | | [Type operation](#type-operation) | [Type assertion](Basic%20Types.md#type-assertions) | |
+| `as` | `import {`_item_ `as` _name_`} from "`_path_`"`<br>`import * as `_name_` from "` _path_ `"` | [Environment/modules](#environment-modules) | [Module import renaming](Modules.md#import) | |
+| `as` | `export as namespace` _namespace_ | [Environment/modules](#environment-modules) | [UMD/isomorphic modules](Modules.md#umd-modules) | |
+| `async` | | [Functions](#functions) | Marks function as asynchronous | [ES draft](http://tc39.github.io/ecmascript-asyncawait/), [PR#1664](https://github.com/Microsoft/TypeScript/issues/1664), [spec](https://github.com/Microsoft/TypeScript/blob/master/doc/spec.md#68-asynchronous-functions) |
+| `await` | | [Functions](#functions) | Waits for value within an asynchronous function | [ES draft](http://tc39.github.io/ecmascript-asyncawait/), [PR#1664](https://github.com/Microsoft/TypeScript/issues/1664), [spec](https://github.com/Microsoft/TypeScript/blob/master/doc/spec.md#68-asynchronous-functions) |
+| `boolean` | | [Predefined type](#predefined-types) | [`boolean` type](Basic%20Types.md#boolean) | |
```
Comment (mhegazy) [inline pages/Keywords.md]: use `"Primitive Type"` instead of `"Predefined type"`. this is how the TypeScript spec refers to them anyways.

Comment (mhegazy) [inline pages/Keywords.md]: this also applies to the rest of the document.

**Comment by mhegazy** [comment]
sorry for the delay, have been out on vacation for a while. a few comments:

- use "Primitive Type" instead of "Predefined type"
- do not think the syntax column is adding much value, it is not filled for most entries, i would just remove it.
- In the first table, description and category are mislabeled. seems like he description column has the category and the category column has the description .

**Comment by zspitz** [comment]
@mhegazy Fixed 'primitive type' to 'predefined type', and headings.
RE: syntax column -- When the same keyword has different meanings (like `abstract` which is applied to both classes and members), the syntax column clarifies the difference between them. This is particularly true when the meanings have little or no relation to each other (e.g. the distinction between `this` function parameters, and `this`-based type guards).
I had originally thought to have the syntax in the first column, but I think the order becomes confusing when the syntax doesn't start with the keyword (`default` vs. `export default` _expr_).

**Comment by metaleap** [comment]
Status on this?  =)

**Comment by zspitz** [comment]
@mhegazy @metaleap There are some changes that need to be made here for 2.1 and 2.2. But if the `release-2.0` branch is obsolete, and this PR needs to be merged somehow into the master branch, then I guess that should be done first.

**Comment by RyanCavanaugh** [comment]
Thank you for contributing to the TypeScript Handbook!

We are currently working through a large backlog of open PRs. As part of this, we're asking authors of PRs with active merge conflicts to assess if their changes are still needed, and if so, open a new PR.

For transparency, we are also working on a full-scale reimagining of the Handbook (see the [TypeScript Roadmap](https://github.com/Microsoft/TypeScript/issues/29288)), meaning we are largely closing down work on the current version of the Handbook. Typo fixes are always appreciated, but any new content should likely wait until the new Handbook is available so that it can be more naturally adapted to that format.

----

Saved reply above; this technically doesn't have merge conflicts but has a ton of unrelated changes in one PR and I'm not comfortable merging it.

**Pull request closed without merging**